`firebase deploy` fails if local emulator is running

<!-- DO NOT DELETE
validate_template=true
template_path=.github/ISSUE_TEMPLATE/---report-a-bug.md
-->


### [REQUIRED] Version info

<!-- What versions of the following libraries are you using? Note that your issue may already
be fixed in the latest versions. -->

**node:** v12.16.2

**firebase-functions:** ^3.6.1

**firebase-tools:** 8.4.3

<!-- Run firebase --version and print the output here-->

**firebase-admin:** ^8.10.0

### [REQUIRED] Test case

Heres a snippet of my npm `scripts`
```json
 {
  "scripts": {
    "build": "tsc",
    "serve": "firebase functions:config:get > .runtimeconfig.json && tsc -w | firebase emulators:start --only functions",
    "deploy": "firebase use prod && firebase deploy --only functions",
  },
}
```

### [REQUIRED] Steps to reproduce

1. Open a terminal, run `npm run serve`.
2. Open another terminal run `npm run deploy`.
3. 2nd terminal outputs "Error: An unexpected error has occurred."

### [REQUIRED] Expected behavior

`firebase deploy` should work while `npm run serve` is still running

### [REQUIRED] Actual behavior

`firebase deploy` errors with no meaningful error message while npm run serve is still running

![](https://i.gyazo.com/d20913cb7abb3b653fbca04f32e99be2.png)

<!-- Please copy and paste any error logs from https://console.firebase.google.com/project/_/functions/logs.
     If you're experiencing a deployment issue, please copy and paste the entirety of firebase-debug.log -->

Also, I'm not able to open `firebase-debug.log`, i get "Access denied" even when trying to open the file in the windows cmd line, opened as Administrator.

### Were you able to successfully deploy your functions?

Yes, but only after i killed the process started by `npm run serve`

<!-- When you ran `firebase deploy`, did you see any error messages? -->

```sh
$ npm run deploy

> functions@ deploy C:\Users\Eric\workspace\pep\optimizers\functions
> firebase use prod && firebase deploy --debug --only functions

Now using alias prod (prod-pep)
[2020-07-16T21:49:08.730Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2020-07-16T21:49:08.732Z] > authorizing via signed-in user
[2020-07-16T21:49:08.732Z] [iam] checking project prod-pep for permissions ["cloudfunctions.functions.create","cloudfunctions.functions.delete","cloudfunctions.functions.get","cloudfunctions.functions.list","cloudfunctions.functions.update","cloudfunctions.operations.get","firebase.projects.get"]
[2020-07-16T21:49:08.746Z] >>> HTTP REQUEST POST https://cloudresourcemanager.googleapis.com/v1/projects/prod-pep:testIamPermissions
 {"permissions":["cloudfunctions.functions.create","cloudfunctions.functions.delete","cloudfunctions.functions.get","cloudfunctions.functions.list","cloudfunctions.functions.update","cloudfunctions.operations.get","firebase.projects.get"]}
[2020-07-16T21:49:08.785Z] Error: EPERM: operation not permitted, stat 'C:\Users\Eric\workspace\pep\optimizers\functions\firebase-debug.log'

Error: An unexpected error has occurred.
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! functions@ deploy: `firebase use prod && firebase deploy --debug --only functions`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the functions@ deploy script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Eric\AppData\Roaming\npm-cache\_logs\2020-07-16T21_49_09_074Z-debug.log
```

--- a/.gitignore
+++ b/.gitignore
@@ -3,6 +3,7 @@ node_modules
 /coverage
 /.nyc_output
 firebase-debug.log
+firebase-debug.*.log
 npm-debug.log
 yarn.lock
 .npmrc

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,3 +4,4 @@
 - Fixes issue where `emulators:exec` could fail to shut down cleanly (#2477).
 - Fixes issue where database emulator did not properly load initial rules (#2483).
 - Allow starting the UI with `emulators:exec` using the `--ui` flag.
+- Fixes issue where multiple CLI instances compete for the same log (#2464).

--- a/scripts/client-integration-tests/test-project/.gitignore
+++ b/scripts/client-integration-tests/test-project/.gitignore
@@ -5,6 +5,7 @@ npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
 firebase-debug.log*
+firebase-debug.*.log*
 
 # Firebase cache
 .firebase/

--- a/src/bin/firebase.js
+++ b/src/bin/firebase.js
@@ -46,15 +46,18 @@ function findAvailableLogFile() {
   for (const c of candidates) {
     const logFilename = path.join(process.cwd(), c);
 
-    if (!fs.existsSync(logFilename)) {
-      return logFilename;
-    }
-
     try {
-      fs.statSync(logFilename);
+      const fd = fs.openSync(logFilename, "r+");
+      fs.closeSync(fd);
       return logFilename;
     } catch (e) {
-      // Continue...
+      if (e.code === "ENOENT") {
+        // File does not exist, which is fine
+        return logFilename;
+      }
+
+      // Any other error (EPERM, etc) means we won't be able to log to
+      // this file so we skip it.
     }
   }
 

--- a/src/bin/firebase.js
+++ b/src/bin/firebase.js
@@ -45,6 +45,11 @@ function findAvailableLogFile() {
 
   for (const c of candidates) {
     const logFilename = path.join(process.cwd(), c);
+
+    if (!fs.existsSync(logFilename)) {
+      return logFilename;
+    }
+
     try {
       fs.statSync(logFilename);
       return logFilename;

--- a/src/bin/firebase.js
+++ b/src/bin/firebase.js
@@ -37,7 +37,26 @@ const handlePreviewToggles = require("../handlePreviewToggles");
 const utils = require("../utils");
 let cmd;
 
-const logFilename = path.join(process.cwd(), "/firebase-debug.log");
+function findAvailableLogFile() {
+  const candidates = ["firebase-debug.log"];
+  for (let i = 1; i < 10; i++) {
+    candidates.push(`firebase-debug.${i}.log`);
+  }
+
+  for (const c of candidates) {
+    const logFilename = path.join(process.cwd(), c);
+    try {
+      fs.statSync(logFilename);
+      return logFilename;
+    } catch (e) {
+      // Continue...
+    }
+  }
+
+  throw new Error("Unable to obtain permissions for firebase-debug.log");
+}
+
+const logFilename = findAvailableLogFile();
 
 if (!process.env.DEBUG && _.includes(args, "--debug")) {
   process.env.DEBUG = true;

--- a/src/listFiles.ts
+++ b/src/listFiles.ts
@@ -5,7 +5,7 @@ export function listFiles(cwd: string, ignore: string[] = []): string[] {
     cwd,
     dot: true,
     follow: true,
-    ignore: ["**/firebase-debug.log", ".firebase/*"].concat(ignore),
+    ignore: ["**/firebase-debug.log", "**/firebase-debug.*.log", ".firebase/*"].concat(ignore),
     nodir: true,
     nosort: true,
   });

--- a/src/prepareFunctionsUpload.js
+++ b/src/prepareFunctionsUpload.js
@@ -64,7 +64,11 @@ var _packageSource = function(options, sourceDir, configValues) {
   // We ignore any CONFIG_DEST_FILE that already exists, and write another one
   // with current config values into the archive in the "end" handler for reader
   var ignore = options.config.get("functions.ignore", ["node_modules", ".git"]);
-  ignore.push("firebase-debug.log", CONFIG_DEST_FILE /* .runtimeconfig.json */);
+  ignore.push(
+    "firebase-debug.log",
+    "firebase-debug.*.log",
+    CONFIG_DEST_FILE /* .runtimeconfig.json */
+  );
   return fsAsync
     .readdirRecursive({ path: sourceDir, ignore: ignore })
     .then(function(files) {

--- a/templates/_gitignore
+++ b/templates/_gitignore
@@ -5,6 +5,7 @@ npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
 firebase-debug.log*
+firebase-debug.*.log*
 
 # Firebase cache
 .firebase/

